datafusion/src/optimizer/projection_push_down.rs

```diff
@@ -417,9 +416,15 @@ fn optimize_plan(
                 })
                 .collect::<Result<Vec<_>>>()?;
 
+            let mut new_schema = Vec::new();
+            for df_field in schema.fields() {
+                if union_required_fields.contains(df_field.field()) {
+                    new_schema.push(df_field.clone());
+                }
+            }
```

Comment:
I don't know if you are into the whole functional programming thing, but you can probably write this without a `mut Vec` like

```rust
            let new_schema = DFSchema::new(
                schema.fields()
                    .iter()
                    .filter(|df_field| {
                        union_required_fields.contains(df_field.field())
                    })
                    .cloned()
                    .collect()
            )?;
```

I don't think it makes any practical difference (what you have here is fine too) I just figured I would point it out

Reply:
I would prefer us to go functional ere to keep the style consistent within the code base :)